File: core/src/model/file_metadata.rs
use crate::model::account::Username;
use crate::model::crypto::{FolderAccessInfo, SignedValue, UserAccessInfo};
use serde::{Deserialize, Serialize};
use std::clone::Clone;
use std::collections::HashMap;
use uuid::Uuid;

#[derive(Clone, PartialEq, Debug, Deserialize, Serialize, Copy)]
pub enum FileType {
    Document,
    Folder,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
pub struct FileMetadata {
    pub id: Uuid,
    pub file_type: FileType,
    pub parent: Uuid,
    pub name: String,
    pub owner: String,
    pub signature: SignedValue,
    pub metadata_version: u64,
    pub content_version: u64,
    pub deleted: bool,
    pub user_access_keys: HashMap<Username, UserAccessInfo>,
    pub folder_access_keys: FolderAccessInfo,
}

File: clients/swift/SwiftLockbookCore/Sources/SwiftLockbookCore/Models/FileMetadata.swift
//
//  FileMetadata.swift
//  ios_client
//
//  Created by Raayan Pillai on 4/11/20.
//  Copyright © 2020 Lockbook. All rights reserved.
//

import Foundation

struct FileMetadata: Codable, Identifiable {
    var fileType: FileType
    var id: UUID
    var parent: UUID
    var name: String
    var owner: String
    var contentVersion: Int
    var metadataVersion: Int
    var deleted: Bool
    var signature: SignedValue = SignedValue(content: "", signature: "")
    var userAccessKeys: [Account.Username : UserAccessInfo] = .init()
    var folderAccessKeys: FolderAccessInfo = FolderAccessInfo(folderId: .init(), accessKey: .init(garbage: "", nonce: ""))
}

enum FileType: String, Codable {
    case Document
    case Folder
}

struct FolderAccessInfo: Codable {
    var folderId: UUID
    var accessKey: EncryptedValueWithNonce
}

struct UserAccessInfo: Codable {
    var username: Account.Username
    var publicKey: RSAPublicKey
    var accessKey: EncryptedValue
}

struct SignedValue: Codable {
    var content: String
    var signature: String
}

struct EncryptedValueWithNonce: Codable {
    var garbage: String
    var nonce: String
}

struct EncryptedValue: Codable {
    var garbage: String
}

struct RSAPublicKey: Codable {
    var n: [UInt]
    var e: [UInt]
}
